### PR TITLE
Signup form input validation

### PR DESCRIPTION
**Pull request by harunugurlu**
- Added input validation for the signup form at ``/signup`` using ``react-hook-form`` and ``zod```. __Validations follow the following rules:__
1. Name should be between A-B amount of characters long and should only include alphabetical characters, hyphens, and apostrophes.
2. Email should match a verified email format
3. Password should follow strict format/recommendations to create secure password
3.  Password & password confirmation should match  
- Adjusted the margins to enhance UI/UX

### CODE REVIEW

File: src/app/signup/page.tsx [hunk 1]
```diff
@@ -35,42 +68,58 @@ const SignUp = () => {
             <h2 className="text-3xl mb-5 font-semibold">Create your account</h2>
           </div>
           <div className="flex items-center px-12 mb-10 w-full">
-            <form action="" className="w-full">
-              <div className="flex mb-10">
+            <form onSubmit={handleSubmit(onSubmit)} action="" className="w-full">
```
Comment (coltonehrman) [inline src/app/signup/page.tsx:71]: I think we can go ahead and remove the `action=""` attribute here now.

File: src/app/signup/page.tsx [hunk 2]
```diff
@@ -1,8 +1,41 @@
...
 import { GradButton } from "@/components/ui/grad-button";
 import SignUpImage from "@/components/ui/signup-image";
+import { z } from "zod";
+import { zodResolver } from "@hookform/resolvers/zod";
+import { useForm } from "react-hook-form";
 
 const SignUp = () => {
+
+  // Example values for A and B
+  const A = 3; // Minimum length
```
Comment (coltonehrman) [inline src/app/signup/page.tsx:13]: Let's name these with some more context. I like `MIN_NAME_LENGTH` and `MAX_NAME_LENGTH`

File: src/app/signup/page.tsx [hunk 3]
```diff
@@ -1,8 +1,41 @@
...
+  const B = 30; // Maximum length
+
+  const signUpSchema = z.object({
+    fullName: z.string()
+      .min(A, { message: `Name must be at least ${A} characters long` })
+      .max(B, { message: `Name must be no more than ${B} characters long` })
+      .regex(/^[a-zA-Z'-]+$/, { message: "Name can only include alphabetical characters, hyphens, and apostrophes" }),
```
Comment (coltonehrman) [inline src/app/signup/page.tsx:20]: What about spaces? If you try to type `Colton Ehrman` will it work?

**Comment by coltonehrman** [comment]
@harunugurlu can you assign your name to the task and also move it in-progress please? Thank you :)

File: src/app/signup/page.tsx [hunk 4]
```diff
@@ -1,8 +1,41 @@
...
 import { GradButton } from "@/components/ui/grad-button";
 import SignUpImage from "@/components/ui/signup-image";
+import { z } from "zod";
+import { zodResolver } from "@hookform/resolvers/zod";
+import { useForm } from "react-hook-form";
 
 const SignUp = () => {
+
+  // Example values for A and B
+  const MIN_NAME_LENGTH = 3; // Minimum length
```
Comment (coltonehrman) [inline src/app/signup/page.tsx:13]: *Nit:* we can remove the comment `// Minimum length` as it is redundant and apparent from the code now :)

File: src/app/signup/page.tsx (hunk 2 above)
Comment (harunugurlu) [reply src/app/signup/page.tsx]: Fixed it

File: src/app/signup/page.tsx (hunk 3 above)
Comment (harunugurlu) [reply src/app/signup/page.tsx]: Fixed it so it allows internal spaces (no spaces at the start or at the end) as well

File: src/app/signup/page.tsx [hunk 5]
```diff
@@ -1,8 +1,41 @@
...
 import Image from "next/image";
 import { GradButton } from "@/components/ui/grad-button";
 import SignUpImage from "@/components/ui/signup-image";
+import { z } from "zod";
+import { zodResolver } from "@hookform/resolvers/zod";
+import { useForm } from "react-hook-form";
 
 const SignUp = () => {
+
+  // Example values for A and B
```
Comment (coltonehrman) [inline src/app/signup/page.tsx:12]: 🪓

File: src/app/signup/page.tsx (hunk 1 above)
Comment (harunugurlu) [reply src/app/signup/page.tsx]: Gone

File: src/app/signup/page.tsx [hunk 6]
```diff
@@ -1,8 +1,41 @@
...
+  const MIN_NAME_LENGTH = 3; // Minimum length
+  const MAX_NAME_LENGTH = 30; // Maximum length
+
+  const signUpSchema = z.object({
+    fullName: z.string()
+      .min(MIN_NAME_LENGTH, { message: `Name must be at least ${MIN_NAME_LENGTH} characters long` })
+      .max(MAX_NAME_LENGTH, { message: `Name must be no more than ${MAX_NAME_LENGTH} characters long` })
+      .regex(/^[a-zA-Z'-][a-zA-Z' -]*[a-zA-Z'-]$/, { message: "Name can only include alphabetical characters, hyphens, apostrophes, and internal spaces" }),
```
Comment (coltonehrman) [inline src/app/signup/page.tsx:20]: This regex seems a bit odd to me as it repeats the pattern 3 times, and the second repetition uses the `*` which should mean "0 or more". Does `/^[a-zA-Z'-]+$/` not work?

Comment (harunugurlu) [reply src/app/signup/page.tsx:20]: ``/^[a-zA-Z'-]+$/`` this doesn't allow spaces. I am simplifying it to ``/^[a-zA-Z'-]+(?: [a-zA-Z'-]+)*$/``

File: src/app/signup/page.tsx (hunk 5 above)
Comment (harunugurlu) [reply src/app/signup/page.tsx:12]: Gone

File: src/app/signup/page.tsx (hunk 4 above)
Comment (harunugurlu) [reply src/app/signup/page.tsx:13]: Yes

**Review by coltonehrman — APPROVED** [review]
LGTM! 🎉

**Pull request merged**